### PR TITLE
[WIP] Post review

### PR DESCRIPTION
**Pull request by pkseverance**
Summary of changes: Reviews can now be posted for a given shelter id. Error handling and testing have yet to be implemented, as I need to read up on error handling in express. I would prefer to wait until we've setup a testing environment to write tests, since it seems like good practice to keep development and testing data  in separate databases. 

This endpoint is a work in progress. Please don't merge yet.

### CODE REVIEW

File: src/controllers/reviews_controller.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+  const review = await prisma.review.create({
+      data: {
+        shelterId: shelterId,
+        cleanliness: cleanliness,
+        safety: safety,
+        staff: staff,
+      },
+    });
+    res.status(204).send();
```
Comment (Tscasady) [inline src/controllers/reviews_controller.ts:23]: I think we should send back the newly created object in the body of the json response.

**Review by Tscasady — CHANGES_REQUESTED** [review]
A good start, just going to wait on tests, see the PR for #4.

File: src/controllers/reviews_controller.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import express from 'express';
+import { PrismaClient, Prisma } from '@prisma/client'
+
+export const reviewController = express.Router();
+
+const prisma = new PrismaClient();
+
```
[the remark this answers is not in the document]
Comment (pkseverance) [reply src/controllers/reviews_controller.ts:7]: Thanks! Will do.

File: src/controllers/reviews_controller.ts (hunk 1 above)
Comment (pkseverance) [reply src/controllers/reviews_controller.ts:23]: Agreed. Thanks!

**Comment by pkseverance** [comment]
Post reviews should now be functional. I implemented basic error handling, did some refactoring, and implemented unit testing for review post requests. Feel free to merge now.

File: package.json [hunk 3]
```diff
@@ -7,7 +7,8 @@
     "dev": "ts-node-dev --respawn --pretty --transpile-only src/server.ts",
     "testLocal": "dotenv -e .env.test npm run test",
     "test": "mocha --file ./test/helper.ts --require ts-node/register test/**/*-test.ts",
-    "start": "tsc && node dist/server.js"
```
Comment (Tscasady) [inline package.json:10]: This start script ought not be removed, it is used in deployment. I assume this was just a merge conflict error.

File: test/requests/reviews/reviewRequest-test.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+      expect(res.body.data.attributes).to.have.key('cleanliness');
+      expect(res.body.data.attributes.cleanliness).to.eq(10);
+
+      expect(res.body.data.attributes).to.have.key('safety');
+      expect(res.body.data.attributes.safety).to.eq(10)
+
+      expect(res.body.data.attributes).to.have.key('staff')
+      expect(res.body.data.attributes.staff).to.eq(9.8)
+
```
Comment (Tscasady) [inline test/requests/reviews/reviewRequest-test.ts:76]: I think we ought to include a test that has some null values, demonstrating that a user does not have to rate all aspects of a shelter.

File: test/requests/reviews/reviewRequest-test.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  it('create a review with given attributes', async () => {
+    const reviewData: ReviewPost = {
+      shelterId: shelter.id,
+      cleanliness: 10,
+      safety: 10,
+      staff: 9.8 
+    }
+
+    chai
```
Comment (Tscasady) [inline test/requests/reviews/reviewRequest-test.ts:47]: All of these tests are giving false positives given the way you have handled async execution. For example I added 
```expect(res.body.data.id).to.equal(12312322) ``` And it passed. In the words of Brian, never trust a passing test! Make sure you get it to fail first.

File: src/serializers/review.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+      type: 'review',
+      attributes: {
+        shelterId: review.shelterId,
+        cleanliness: review.cleanliness,
+        safety: review.safety,
+        staff: review.staff,
+      },
+    }
+  } else {
+    return { data: {} }
```
Comment (Tscasady) [inline src/serializers/review.ts:16]: This implementation was a good placeholder for 'errors' but I think now we are in a position to return real 404 errors and validation errors instead of simply an empty object. See #20 and #21

**Review by Tscasady — CHANGES_REQUESTED** [review]
I left some notes for how to proceed with error handling / validation, an example was put up today in other PRs. I'm mostly concerned with testing, your current test file is giving false positives.

File: src/serializers/review.ts (hunk 6 above)
Comment (Tscasady) [reply src/serializers/review.ts:16]: If you would like to make a separate issue and PR for handling errors that is fine with me.

**Comment by pkseverance** [comment]
Everything should work now.

File: package.json [hunk 7]
```diff
@@ -6,8 +6,11 @@
   "scripts": {
     "dev": "ts-node-dev --respawn --pretty --transpile-only src/server.ts",
     "testLocal": "dotenv -e .env.test npm run test",
-    "test": "mocha --file ./test/helper.ts --require ts-node/register 'test/**/*-test.ts'",
-    "start": "tsc && node dist/server.js"
+    "test": "mocha --file ./test/helper.ts --require ts-node/register test/**/*-test.ts",
```
Comment (Tscasady) [inline package.json:9]: you've removed the ' quotes around the 'test/**/*-test.ts' and it is no longer properly running all tests. Your most recent github actions job only ran two tests.

File: test/requests/reviews/reviewRequest-test.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+type ShelterPost = {
+  name: string
+  streetAddress: string
+  state: string
+  zip: number
+  phoneNumber: string
+}
+
```
Comment (Tscasady) [inline test/requests/reviews/reviewRequest-test.ts:22]: These type declarations are now available in shelter.types.ts and review.types.ts

File: test/requests/shelters/shelterRequest-test.ts [hunk 9]
```diff
@@ -69,6 +77,56 @@ describe('Post Shelter', () => {
...
+  it('create a shelter given a name and address and phone number AND a websiteurl', (done) => {
+    const shelterParams: ShelterPost = {
+      name: 'Golden Sun',
+      streetAddress: '1234 Black St',
+      state: 'NY',
+      zip: 78123,
+      phoneNumber: '2134568765',
+      websiteUrl: "www.fake.com"
+    }
+
```
Comment (Tscasady) [inline test/requests/shelters/shelterRequest-test.ts:103]: This file has had some significant changes, and the version on main is correct. This file is not related to this PR and is out of date.

File: src/serializers/review.ts (hunk 6 above)
Comment (pkseverance) [reply src/serializers/review.ts:16]: I'm going to work on error handling separately.

**Review by Tscasady — APPROVED** [review]
Looks like you got all the testing sorted, and the CI checks look correct. Make sure you add an issue/card to project board for error/sad path for POST review.
